### ISSUE TITLE
NameError: uninitialized constant ActiveRecord::ConnectionAdapters::DatabaseStatements::TransactionManager when calling reset_transaction

### ISSUE THREAD

**Issue by synthead** [rails/rails#26441]
Saw this exception pop up.  I'm not sure how or why it happened, so I can't provide steps to reproduce it, but I'm happy to try anything to help.

```
2016-09-09T18:12:08.225Z 8 TID-ouaqkszjs WARN: NameError: uninitialized constant ActiveRecord::ConnectionAdapters::DatabaseStatements::TransactionManager
Did you mean?  ActiveRecord::ConnectionAdapters::TransactionManager
               ActiveRecord::ConnectionAdapters::Transaction
               ActiveRecord::ConnectionAdapters::TransactionState
2016-09-09T18:12:08.236Z 8 TID-ouaqkszjs WARN: /usr/lib/ruby/gems/2.3.0/gems/activerecord-5.0.0.1/lib/active_record/connection_adapters/abstract/database_statements.rb:247:in `reset_transaction'
/usr/lib/ruby/gems/2.3.0/gems/activerecord-5.0.0.1/lib/active_record/connection_adapters/abstract/database_statements.rb:6:in `initialize'
/usr/lib/ruby/gems/2.3.0/gems/activerecord-5.0.0.1/lib/active_record/connection_adapters/abstract/query_cache.rb:24:in `initialize'
/usr/lib/ruby/gems/2.3.0/gems/activerecord-5.0.0.1/lib/active_record/connection_adapters/abstract_adapter.rb:99:in `initialize'
/usr/lib/ruby/gems/2.3.0/gems/activerecord-5.0.0.1/lib/active_record/connection_adapters/postgresql_adapter.rb:209:in `initialize'
/usr/lib/ruby/gems/2.3.0/gems/activerecord-5.0.0.1/lib/active_record/connection_adapters/postgresql_adapter.rb:37:in `new'
/usr/lib/ruby/gems/2.3.0/gems/activerecord-5.0.0.1/lib/active_record/connection_adapters/postgresql_adapter.rb:37:in `postgresql_connection'
/usr/lib/ruby/gems/2.3.0/gems/activerecord-5.0.0.1/lib/active_record/connection_adapters/abstract/connection_pool.rb:721:in `new_connection'
/usr/lib/ruby/gems/2.3.0/gems/activerecord-5.0.0.1/lib/active_record/connection_adapters/abstract/connection_pool.rb:765:in `checkout_new_connection'
/usr/lib/ruby/gems/2.3.0/gems/activerecord-5.0.0.1/lib/active_record/connection_adapters/abstract/connection_pool.rb:744:in `try_to_checkout_new_connection'
/usr/lib/ruby/gems/2.3.0/gems/activerecord-5.0.0.1/lib/active_record/connection_adapters/abstract/connection_pool.rb:705:in `acquire_connection'
/usr/lib/ruby/gems/2.3.0/gems/activerecord-5.0.0.1/lib/active_record/connection_adapters/abstract/connection_pool.rb:501:in `checkout'
/usr/lib/ruby/gems/2.3.0/gems/activerecord-5.0.0.1/lib/active_record/connection_adapters/abstract/connection_pool.rb:364:in `connection'
/usr/lib/ruby/gems/2.3.0/gems/activerecord-5.0.0.1/lib/active_record/connection_adapters/abstract/connection_pool.rb:875:in `retrieve_connection'
/usr/lib/ruby/gems/2.3.0/gems/activerecord-5.0.0.1/lib/active_record/connection_handling.rb:128:in `retrieve_connection'
/usr/lib/ruby/gems/2.3.0/gems/activerecord-5.0.0.1/lib/active_record/connection_handling.rb:91:in `connection'
/usr/lib/ruby/gems/2.3.0/gems/activerecord-5.0.0.1/lib/active_record/query_cache.rb:47:in `block in install_executor_hooks'
/usr/lib/ruby/gems/2.3.0/gems/activesupport-5.0.0.1/lib/active_support/callbacks.rb:396:in `instance_exec'
/usr/lib/ruby/gems/2.3.0/gems/activesupport-5.0.0.1/lib/active_support/callbacks.rb:396:in `block in make_lambda'
/usr/lib/ruby/gems/2.3.0/gems/activesupport-5.0.0.1/lib/active_support/callbacks.rb:169:in `block (2 levels) in halting'
/usr/lib/ruby/gems/2.3.0/gems/activesupport-5.0.0.1/lib/active_support/callbacks.rb:547:in `block (2 levels) in default_terminator'
/usr/lib/ruby/gems/2.3.0/gems/activesupport-5.0.0.1/lib/active_support/callbacks.rb:546:in `catch'
/usr/lib/ruby/gems/2.3.0/gems/activesupport-5.0.0.1/lib/active_support/callbacks.rb:546:in `block in default_terminator'
/usr/lib/ruby/gems/2.3.0/gems/activesupport-5.0.0.1/lib/active_support/callbacks.rb:170:in `block in halting'
/usr/lib/ruby/gems/2.3.0/gems/activesupport-5.0.0.1/lib/active_support/callbacks.rb:454:in `block in call'
/usr/lib/ruby/gems/2.3.0/gems/activesupport-5.0.0.1/lib/active_support/callbacks.rb:454:in `each'
/usr/lib/ruby/gems/2.3.0/gems/activesupport-5.0.0.1/lib/active_support/callbacks.rb:454:in `call'
/usr/lib/ruby/gems/2.3.0/gems/activesupport-5.0.0.1/lib/active_support/callbacks.rb:101:in `__run_callbacks__'
/usr/lib/ruby/gems/2.3.0/gems/activesupport-5.0.0.1/lib/active_support/callbacks.rb:750:in `_run_complete_callbacks'
/usr/lib/ruby/gems/2.3.0/gems/activesupport-5.0.0.1/lib/active_support/callbacks.rb:90:in `run_callbacks'
/usr/lib/ruby/gems/2.3.0/gems/activesupport-5.0.0.1/lib/active_support/execution_wrapper.rb:107:in `complete!'
/usr/lib/ruby/gems/2.3.0/gems/activesupport-5.0.0.1/lib/active_support/execution_wrapper.rb:64:in `ensure in block in run!'
/usr/lib/ruby/gems/2.3.0/gems/activesupport-5.0.0.1/lib/active_support/execution_wrapper.rb:64:in `block in run!'
/usr/lib/ruby/gems/2.3.0/gems/activesupport-5.0.0.1/lib/active_support/execution_wrapper.rb:58:in `tap'
/usr/lib/ruby/gems/2.3.0/gems/activesupport-5.0.0.1/lib/active_support/execution_wrapper.rb:58:in `run!'
/usr/lib/ruby/gems/2.3.0/gems/activesupport-5.0.0.1/lib/active_support/execution_wrapper.rb:74:in `wrap'
/usr/lib/ruby/gems/2.3.0/gems/activesupport-5.0.0.1/lib/active_support/reloader.rb:67:in `wrap'
/usr/lib/ruby/gems/2.3.0/gems/activejob-5.0.0.1/lib/active_job/railtie.rb:25:in `block (3 levels) in <class:Railtie>'
/usr/lib/ruby/gems/2.3.0/gems/activesupport-5.0.0.1/lib/active_support/callbacks.rb:391:in `instance_exec'
/usr/lib/ruby/gems/2.3.0/gems/activesupport-5.0.0.1/lib/active_support/callbacks.rb:391:in `block in make_lambda'
/usr/lib/ruby/gems/2.3.0/gems/activesupport-5.0.0.1/lib/active_support/callbacks.rb:285:in `block in halting'
/usr/lib/ruby/gems/2.3.0/gems/activesupport-5.0.0.1/lib/active_support/callbacks.rb:447:in `block in around'
/usr/lib/ruby/gems/2.3.0/gems/activesupport-5.0.0.1/lib/active_support/callbacks.rb:455:in `call'
/usr/lib/ruby/gems/2.3.0/gems/activesupport-5.0.0.1/lib/active_support/callbacks.rb:101:in `__run_callbacks__'
/usr/lib/ruby/gems/2.3.0/gems/activesupport-5.0.0.1/lib/active_support/callbacks.rb:750:in `_run_execute_callbacks'
/usr/lib/ruby/gems/2.3.0/gems/activesupport-5.0.0.1/lib/active_support/callbacks.rb:90:in `run_callbacks'
/usr/lib/ruby/gems/2.3.0/gems/activejob-5.0.0.1/lib/active_job/execution.rb:20:in `execute'
/usr/lib/ruby/gems/2.3.0/gems/activejob-5.0.0.1/lib/active_job/queue_adapters/sidekiq_adapter.rb:40:in `perform'
/usr/lib/ruby/gems/2.3.0/gems/sidekiq-4.1.2/lib/sidekiq/processor.rb:152:in `execute_job'
/usr/lib/ruby/gems/2.3.0/gems/sidekiq-4.1.2/lib/sidekiq/processor.rb:134:in `block (2 levels) in process'
/usr/lib/ruby/gems/2.3.0/gems/sidekiq-4.1.2/lib/sidekiq/middleware/chain.rb:128:in `block in invoke'
/usr/lib/ruby/gems/2.3.0/gems/sidekiq-4.1.2/lib/sidekiq/middleware/server/active_record.rb:6:in `call'
/usr/lib/ruby/gems/2.3.0/gems/sidekiq-4.1.2/lib/sidekiq/middleware/chain.rb:130:in `block in invoke'
/usr/lib/ruby/gems/2.3.0/gems/sidekiq-4.1.2/lib/sidekiq/middleware/server/logging.rb:11:in `block in call'
/usr/lib/ruby/gems/2.3.0/gems/sidekiq-4.1.2/lib/sidekiq/logging.rb:32:in `with_context'
/usr/lib/ruby/gems/2.3.0/gems/sidekiq-4.1.2/lib/sidekiq/middleware/server/logging.rb:7:in `call'
/usr/lib/ruby/gems/2.3.0/gems/sidekiq-4.1.2/lib/sidekiq/middleware/chain.rb:130:in `block in invoke'
/usr/lib/ruby/gems/2.3.0/gems/sidekiq-4.1.2/lib/sidekiq/middleware/chain.rb:133:in `invoke'
/usr/lib/ruby/gems/2.3.0/gems/sidekiq-4.1.2/lib/sidekiq/processor.rb:129:in `block in process'
/usr/lib/ruby/gems/2.3.0/gems/sidekiq-4.1.2/lib/sidekiq/processor.rb:168:in `stats'
/usr/lib/ruby/gems/2.3.0/gems/sidekiq-4.1.2/lib/sidekiq/processor.rb:128:in `process'
/usr/lib/ruby/gems/2.3.0/gems/sidekiq-4.1.2/lib/sidekiq/processor.rb:80:in `process_one'
/usr/lib/ruby/gems/2.3.0/gems/sidekiq-4.1.2/lib/sidekiq/processor.rb:68:in `run'
/usr/lib/ruby/gems/2.3.0/gems/sidekiq-4.1.2/lib/sidekiq/util.rb:17:in `watchdog'
/usr/lib/ruby/gems/2.3.0/gems/sidekiq-4.1.2/lib/sidekiq/util.rb:25:in `block in safe_thread'
```

Something tells me that [this line](https://github.com/rails/rails/blob/b27acd0b69a093e1d9d91ddf52c9f6e1289e28e0/activerecord/lib/active_record/connection_adapters/abstract/database_statements.rb#L248)...
```ruby
      def reset_transaction #:nodoc:
        @transaction_manager = TransactionManager.new(self)
      end
```
...should probably be updated to...
```ruby
      def reset_transaction #:nodoc:
        @transaction_manager = ActiveRecord::ConnectionAdapters::TransactionManager.new(self)
      end
```

Also, it looks like the [two](https://github.com/rails/rails/blob/3fc0bbf008f0e935ab56559f119c9ea8250bfddd/activerecord/test/cases/transactions_test.rb#L659-L671) [tests](https://github.com/rails/rails/blob/3fc0bbf008f0e935ab56559f119c9ea8250bfddd/activerecord/test/cases/transactions_test.rb#L673-L685) actually call `TransactionManager` explicitly, so I believe that the `reset_transaction` method doesn't have a test to catch the above exception:

```ruby
  def test_transactions_state_from_rollback
    connection = Topic.connection
    transaction = ActiveRecord::ConnectionAdapters::TransactionManager.new(connection).begin_transaction

    assert transaction.open?
    assert !transaction.state.rolledback?
    assert !transaction.state.committed?

    transaction.rollback

    assert transaction.state.rolledback?
    assert !transaction.state.committed?
  end

  def test_transactions_state_from_commit
    connection = Topic.connection
    transaction = ActiveRecord::ConnectionAdapters::TransactionManager.new(connection).begin_transaction

    assert transaction.open?
    assert !transaction.state.rolledback?
    assert !transaction.state.committed?

    transaction.commit

    assert !transaction.state.rolledback?
    assert transaction.state.committed?
  end
```

### System configuration
**Rails version**: 5.0.0.1
**Ruby version**: 2.3.1p112

**Comment by matthewd** [comment]
#26273 may be related

**Comment by arthurnn** [comment]
@matthewd is this related to code reload?

**Comment by synthead** [comment]
If it helps to mention it, this happened in production mode.

**Comment by arthurnn** [comment]
@synthead does your app has any special `require` design? or caching on the load paths? or anything that could affect this?
I am trying to think why this happens there and not in the other runs.

**Comment by matthewd** [comment]
@arthurnn it *shouldn't* be -- these classes don't get unloaded. But given that something is clearly doing the wrong thing, I wouldn't rule it out.

**Comment by synthead** [comment]
This is in `config/application.rb`:
```ruby
module MyApp
  class Application < Rails::Application
    # Settings in config/environments/* take precedence over those specified here.
    # Application configuration should go into files in config/initializers
    # -- all .rb files in that directory are automatically loaded.

    # Set Time.zone default to the specified zone and make Active Record auto-convert to this zone.
    # Run "rake -D time" for a list of tasks for finding time zone names. Default is UTC.
    # config.time_zone = 'Central Time (US & Canada)'

    # The default locale is :en and all translations from config/locales/*.rb,yml are auto loaded.
    # config.i18n.load_path += Dir[Rails.root.join('my', 'locales', '*.{rb,yml}').to_s]
    # config.i18n.default_locale = :de

    # Do not swallow errors in after_commit/after_rollback callbacks.
    config.active_record.raise_in_transactional_callbacks = true

    # Include ActiveJob concerns when application is eager-loaded.
    config.eager_load_paths.unshift(config.root.join("app/jobs/concerns").to_s)

    # Allow WebSocket requests from any origin.
    config.action_cable.disable_request_forgery_protection = true

    config.after_initialize do
      Rails.application.eager_load!

      ::ACTIONS_MAP = Hash.new.tap do |my_app_actions|
        ActiveJob::Base.descendants.each do |descendant|
          if descendant.included_modules.include?(MyApp)
            action = descendant.name.underscore.chomp("_job")
            my_app_actions[action] = descendant
          end
        end
      end
    end
  end
end
```

Think this might be causing issues?  The strange thing is that it happened unexpectedly about a day after a number of Sidekiq jobs went though.

**Comment by arthurnn** [comment]
I added the correct namespace, which should fix this issue. see https://github.com/rails/rails/commit/f62451a50b2c9119adce7acc53ce3dfffc4d41d5.
I will backport that to 5-0-stable, so it should be available in the next 5.0 release

**Comment by synthead** [comment]
Awesome! Thanks so much!
